### PR TITLE
revritten jwt validation

### PR DESCRIPTION
**Pull request by tymak**
Refactor validation to fit current JWT format, added option for validating as middleware or standalone function

### CODE REVIEW

File: src/models/jwtPermissions/index.js [hunk 1]
```diff
@@ -136,3 +155,48 @@ exports.init = async function ({ publicKeyUrl = _publicKeyUrl, loadPublicKeyFrom
...
+    }
+};
+
+/**
+ *
+ * @param ctx
+ * @param next
+ * @returns {Promise<void>}
+ */
+exports.validateMerchantMidleware = async (ctx, next) => {
```
Comment (VaclavObornik) [inline src/models/jwtPermissions/index.js:177]: typo

File: src/models/jwtPermissions/index.js [hunk 2]
```diff
@@ -136,3 +155,48 @@ exports.init = async function ({ publicKeyUrl = _publicKeyUrl, loadPublicKeyFrom
...
+
+/**
+ *
+ * @param {string}token
+ * @param {string}merchant
+ * @param {number[]|number}permissions
+ * @param { {string|undefined} }
+ * @returns {Promise<void>}
+ */
+exports.authorizeUser = async function (token, merchant, permissions, { publicKeyUrl = _publicKeyUrl } = {}) {
```
Comment (VaclavObornik) [inline src/models/jwtPermissions/index.js:194]: merchant -> merchantId

File: src/models/jwtPermissions/index.js [hunk 3]
```diff
@@ -136,3 +155,48 @@ exports.init = async function ({ publicKeyUrl = _publicKeyUrl, loadPublicKeyFrom
...
+    const permisisonScope = getScope(jwtPayload, PERMISSION_SCOPE);
+    const tokenMerchantId = permisisonScope[2].merchantId;
+    const uriMerchantId = ctx.params.merchantId;
+    validateMerchant(tokenMerchantId, uriMerchantId);
+    await next();
+};
+
+/**
+ *
+ * @param {string}token
```
Comment (VaclavObornik) [inline src/models/jwtPermissions/index.js:188]: fix the formatting, spaces before parameter names in the docs

File: src/models/jwtPermissions/index.js [hunk 4]
```diff
@@ -136,3 +155,48 @@ exports.init = async function ({ publicKeyUrl = _publicKeyUrl, loadPublicKeyFrom
...
+ * @param {string}merchant
+ * @param {number[]|number}permissions
+ * @param { {string|undefined} }
+ * @returns {Promise<void>}
+ */
+exports.authorizeUser = async function (token, merchant, permissions, { publicKeyUrl = _publicKeyUrl } = {}) {
+    const payload = await validateJwt(token, publicKeyUrl);
+    const permisisonScope = getScope(payload, PERMISSION_SCOPE);
```
Comment (VaclavObornik) [inline src/models/jwtPermissions/index.js:196]: typo - try to set your IDE to let you know about the typos, see the green line under the wrong word:
image

File: src/models/jwtPermissions/index.js [hunk 5]
```diff
@@ -136,3 +155,48 @@ exports.init = async function ({ publicKeyUrl = _publicKeyUrl, loadPublicKeyFrom
...
 };
+
+/**
+ *
+ * @param {string}tokenMerchantId
+ * @param {string}uriMerchantId
+ * @returns {void}
+ */
+const validateMerchant = (tokenMerchantId, uriMerchantId) => {
+    if (tokenMerchantId !== uriMerchantId && tokenMerchantId !== '*') {
```
Comment (VaclavObornik) [inline src/models/jwtPermissions/index.js:166]: we should also validate placeId if it is required by the app
(the route has placeId parameter or a placeId is provided by argument authorizeUser function)

in that case we should validate the placeIds in the token includes the id or '*'

File: src/models/jwtPermissions/index.js [hunk 6]
```diff
@@ -136,3 +155,48 @@ exports.init = async function ({ publicKeyUrl = _publicKeyUrl, loadPublicKeyFrom
...
+ *
+ * @param ctx
+ * @param next
+ * @returns {Promise<void>}
+ */
+exports.validateMerchantMidleware = async (ctx, next) => {
+    const { jwtPayload } = ctx.state.jwtPayload;
+    const permisisonScope = getScope(jwtPayload, PERMISSION_SCOPE);
+    const tokenMerchantId = permisisonScope[2].merchantId;
+    const uriMerchantId = ctx.params.merchantId;
```
Comment (VaclavObornik) [inline src/models/jwtPermissions/index.js:181]: nice to have - if the uri contains "merchantPlaceId" parameter, we can split it by '-' and get merchantId and placeId to validate. The merchantPlaceId is commonly used in our APIs, so the middleware would be even more usable

**Review by VaclavObornik — CHANGES_REQUESTED** [review]

File: src/models/jwtPermissions/index.js (hunk 1 above)
Comment (tymak) [reply src/models/jwtPermissions/index.js:177]: done

File: src/models/jwtPermissions/index.js (hunk 2 above)
Comment (tymak) [reply src/models/jwtPermissions/index.js:194]: done

File: src/models/jwtPermissions/index.js (hunk 3 above)
Comment (tymak) [reply src/models/jwtPermissions/index.js:188]: done

Comment (tymak) [reply src/models/jwtPermissions/index.js:188]: done

File: src/models/jwtPermissions/index.js (hunk 4 above)
Comment (tymak) [reply src/models/jwtPermissions/index.js:196]: done

File: src/models/jwtPermissions/index.js (hunk 6 above)
Comment (tymak) [reply src/models/jwtPermissions/index.js:181]: done

File: src/models/jwtPermissions/index.js [hunk 7]
```diff
@@ -158,45 +158,70 @@ exports.init = async function ({ publicKeyUrl = _publicKeyUrl, loadPublicKeyFrom
...
+    const permissionScope = getScope(jwtPayload, PERMISSION_SCOPE);
+    const { merchantId, placesIds } = permissionScope[2];
+    let uriMerchantId;
+    if (ctx.params.merchantPlaceId) {
+        const parsedMerchantPlaceId = ctx.params.merchantPlaceId.split('-');
+        let uriPlacesIds;
+        [uriMerchantId, uriPlacesIds] = parsedMerchantPlaceId;
+        validatePlace(placesIds, uriPlacesIds);
+    } else {
+        uriMerchantId = ctx.params.merchantId;
```
Comment (VaclavObornik) [inline src/models/jwtPermissions/index.js:199]: what if there is also placeId as separate parameter in the URI?

File: src/models/jwtPermissions/index.js [hunk 8]
```diff
@@ -109,18 +109,18 @@ const validatePermissionRightsStrict = (tokenPermissions, permissions) => {
...
+    checkPermissions(tokenPermissions, permissions);
     await next();
 };
 /**
  *
  * @param {string} tokenPermissions
  * @param {number[]|number} permissions
  */
-const validatePermissionRights = (tokenPermissions, permissions) => {
+const checkAtLeastOnePermissions = (tokenPermissions, permissions) => {
```
Comment (VaclavObornik) [inline src/models/jwtPermissions/index.js:123]: grammar: checkAtLeastOnePermissions -> checkAtLeastOnePermission

**Review by VaclavObornik — CHANGES_REQUESTED** [review]

File: src/models/jwtPermissions/index.js [hunk 9]
```diff
@@ -132,10 +132,10 @@ const checkAtLeastOnePermissions = (tokenPermissions, permissions) => {
  * @param {number|number[]} permissions
  * @returns {(function(*, *): Promise<void>)|*}
  */
-exports.checkAtLeastOnePermissionsMiddleWare = (permissions) => async (ctx, next) => {
+exports.checkAtLeastOnePermissionMiddleWare = (permissions) => async (ctx, next) => {
     permissions = typeof permissions !== 'object' ? [permissions] : permissions;
     const tokenPermissions = getScope(ctx.state.jwtPayload, PERMISSION_SCOPE)[1];
```
Comment (VaclavObornik) [inline src/models/jwtPermissions/index.js:137]: if there is no matching scope, an unexpected error will be thrown ("TypeError: Cannot read properties of null")

File: src/models/jwtPermissions/index.js [hunk 10]
```diff
@@ -120,7 +120,7 @@ exports.checkPermissionsMiddleWare = (permissions) => async (ctx, next) => {
  * @param {string} tokenPermissions
  * @param {number[]|number} permissions
  */
-const checkAtLeastOnePermissions = (tokenPermissions, permissions) => {
+const checkAtLeastOnePermission = (tokenPermissions, permissions) => {
     const decodedPermissions = decodePermissions(tokenPermissions);
     const validPermission = permissions.find((i) => decodedPermissions[i]);
```
Comment (VaclavObornik) [inline src/models/jwtPermissions/index.js:125]: this will not work for permissions with index 0

File: src/models/jwtPermissions/index.js [hunk 11]
```diff
@@ -136,3 +155,74 @@ exports.init = async function ({ publicKeyUrl = _publicKeyUrl, loadPublicKeyFrom
...
+/**
+ *
+ * @param ctx
+ * @param next
+ * @returns {Promise<void>}
+ */
+exports.validateMerchantMiddleware = async (ctx, next) => {
+    const { jwtPayload } = ctx.state;
+    const permissionScope = getScope(jwtPayload, PERMISSION_SCOPE);
+    const { merchantId, placesIds } = permissionScope[2];
```
Comment (VaclavObornik) [inline src/models/jwtPermissions/index.js:191]: if the token does not contain the 'perms' scope, the code will fail

File: src/models/jwtPermissions/index.js [hunk 12]
```diff
@@ -136,3 +155,74 @@ exports.init = async function ({ publicKeyUrl = _publicKeyUrl, loadPublicKeyFrom
...
+ * @param {number[]|number|undefined} permissions
+ * @returns {Promise<void>}
+ */
+exports.authorizeUser = async function (
+    token, merchantId, { publicKeyUrl = _publicKeyUrl, placeId = null, permissions } = {}
+) {
+    const payload = await validateJwt(token, publicKeyUrl);
+    const permissionScope = getScope(payload, PERMISSION_SCOPE);
+    const tokenPermissions = permissionScope[1];
+    const { merchantId: tokenMerchantId, placesIds } = permissionScope[2];
```
Comment (VaclavObornik) [inline src/models/jwtPermissions/index.js:224]: if the token does not contain the 'perms' scope, the code will fail

File: src/models/jwtPermissions/index.js [hunk 13]
```diff
@@ -136,3 +155,74 @@ exports.init = async function ({ publicKeyUrl = _publicKeyUrl, loadPublicKeyFrom
...
+    if (placeId) validatePlace(placesIds, placeId);
```
Comment (VaclavObornik) [inline src/models/jwtPermissions/index.js:225]: use blocks for conditions, please

File: src/models/jwtPermissions/index.js [hunk 14]
```diff
@@ -136,3 +155,74 @@ exports.init = async function ({ publicKeyUrl = _publicKeyUrl, loadPublicKeyFrom
...
+    validateMerchant(tokenMerchantId, merchantId);
+    if (permissions) checkAtLeastOnePermission(tokenPermissions, permissions);
```
Comment (VaclavObornik) [inline src/models/jwtPermissions/index.js:227]: use blocks for conditions, please

**Review by VaclavObornik — CHANGES_REQUESTED** [review]

File: src/models/jwtPermissions/index.js (hunk 10 above)
Comment (tymak) [reply src/models/jwtPermissions/index.js:125]: done

File: src/models/jwtPermissions/index.js (hunk 11 above)
Comment (tymak) [reply src/models/jwtPermissions/index.js:191]: done

File: src/models/jwtPermissions/index.js (hunk 12 above)
Comment (tymak) [reply src/models/jwtPermissions/index.js:224]: done

File: src/models/jwtPermissions/index.js (hunk 13 above)
Comment (tymak) [reply src/models/jwtPermissions/index.js:225]: done

File: src/models/jwtPermissions/index.js (hunk 14 above)
Comment (tymak) [reply src/models/jwtPermissions/index.js:227]: done

File: src/models/jwtPermissions/index.js (hunk 9 above)
Comment (tymak) [reply src/models/jwtPermissions/index.js:137]: done

**Review by VaclavObornik — DISMISSED** [review]

**Review by VaclavObornik — APPROVED** [review]

**Pull request merged**